compiler/plugins/target/AMD-AIE/iree-amd-aie/Target/AIETarget.cpp

```diff
@@ -164,131 +205,195 @@ LogicalResult AIETargetBackend::serializeExecutable(
+    // we use the 42 first characters for the kernel name. We suffix ordinal
+    // number to the name to gaurantee uniqueness within the executable.
+    std::string entryPointName = llvm::join_items(
+        "_", exportOpName.substr(0, 42), std::to_string(ordinal));
+    entryPointNames.emplace_back(entryPointName);
+    entryPointOrdinals[entryPointName] = ordinal;
   }
+  uint64_t ordinalCount = entryPointOrdinals.size();
 
-  if (entryPointNames.size() != 1) {
-    return moduleOp.emitOpError("Expected a single entry point");
+  if (entryPointNames.empty()) {
+    return moduleOp.emitOpError("should contain some entry points");
   }
 
-  SmallVector<StringRef> cmdArgs{aie2xclbin,
-                                 inputMlirPath,
-                                 "--peano",
-                                 options.peanoInstallDir,
-                                 "--xclbin-name",
-                                 xclbinPath,
-                                 "--npu-insts-name",
-                                 npuInstPath,
-                                 "--xclbin-kernel-name",
-                                 entryPointNames[0],
-                                 "--tmpdir",
-                                 workDir};
-
-  auto addOpt = [&](StringRef arg, bool value) {
-    if (value) cmdArgs.push_back(arg);
-  };
-  addOpt("--use-chess", options.useChess);
-  addOpt("-v", options.showInvokedCommands);
-  addOpt("--print-ir-after-all", options.aie2xclbinPrintIrAfterAll);
-  addOpt("--print-ir-before-all", options.aie2xclbinPrintIrBeforeAll);
-  addOpt("--disable-threading", options.aie2xclbinDisableTheading);
-  addOpt("--print-ir-module-scope", options.aie2xclbinPrintIrModuleScope);
-  addOpt("--timing", options.aie2xclbinTiming);
-
-  SmallVector<std::string> cmdEnv{};
-  if (options.useChess) {
-    std::string newVitis = "VITIS=" + options.vitisInstallDir;
-    cmdEnv.push_back(newVitis);
-  }
-
-  if (const char *originalPath = ::getenv("PATH")) {
-    std::string newPath = "PATH=" + std::string(originalPath);
-    cmdEnv.push_back(newPath);
-  }
+  SmallString<128> aie2xclbin(options.mlirAieInstallDir);
+  llvm::sys::path::append(aie2xclbin, "bin", "aie2xclbin");
+  std::vector<uint32_t> npuInstrs;
+  std::unique_ptr<llvm::MemoryBuffer> xclbinIn;
 
-  // Chess (if used) will look here for the AIEbuild license.
-  if (const char *originalHome = ::getenv("HOME")) {
-    std::string newHome = std::string("HOME=") + std::string(originalHome);
-    cmdEnv.push_back(newHome);
-  }
+  FlatbufferBuilder builder;
+  iree_amd_aie_hal_xrt_ExecutableDef_start_as_root(builder);
+  SmallVector<iree_amd_aie_hal_xrt_XclbinDef_ref_t> xclbinRefs;
+  SmallVector<iree_amd_aie_hal_xrt_AsmInstDef_ref_t> asmInstrRefs;
+
+  // Per entry-point data.
+  // Note that the following vectors should all be of the same size and
+  // element at index #i is for entry point with ordinal #i!
+  SmallVector<std::string> entryPointNamesFb;
```

Comment:
nit

```suggestion
  SmallVector<std::string> entryPointNamesFb (ordinalCount);
```

etc rather than resize